File: src/main/java/cn/keevol/keenotes/mobilefx/PendingNoteService.java
package cn.keevol.keenotes.mobilefx;

import javafx.application.Platform;
import javafx.beans.property.IntegerProperty;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.concurrent.*;
import java.util.logging.Logger;

/**
 * 离线暂存笔记的调度服务
 * - 暂存发送失败的笔记
 * - 定时重试发送（30分钟间隔）
 * - WebSocket重连时立即触发重试
 */
public class PendingNoteService {
    private static final Logger logger = Logger.getLogger(PendingNoteService.class.getName());
    private static final DateTimeFormatter TS_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    private static final long RETRY_INTERVAL_MINUTES = 30;

    private static PendingNoteService instance;

    private final LocalCacheService localCache;
    private ScheduledExecutorService retryScheduler;
    private volatile boolean retrying = false;

    private PendingNoteService() {
        this.localCache = LocalCacheService.getInstance();
    }

    public static synchronized PendingNoteService getInstance() {
        if (instance == null) {
            instance = new PendingNoteService();
        }
        return instance;
    }

    /**
     * 获取 pending count 的 reactive property（代理到 LocalCacheService）
     */
    public IntegerProperty pendingCountProperty() {
        return localCache.pendingNoteCountProperty();
    }

    /**
     * 暂存一条笔记到本地
     */
    public void savePendingNote(String content, String channel) {
        try {
            localCache.insertPendingNote(content, channel);
            logger.info("Note saved to pending: " + content.substring(0, Math.min(20, content.length())) + "...");
        } catch (Exception e) {
            logger.warning("Failed to save pending note: " + e.getMessage());
        }
    }

    /**
     * 获取所有待发送笔记
     */
    public List<LocalCacheService.PendingNoteData> getPendingNotes() {
        return localCache.getPendingNotes();
    }

    /**
     * 启动定时重试调度器
     */
    public void startRetryScheduler() {
        if (retryScheduler != null && !retryScheduler.isShutdown()) {
            return;
        }
        retryScheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "pending-note-retry");
            t.setDaemon(true);
            return t;
        });
        retryScheduler.scheduleAtFixedRate(
                this::retryPendingNotes,
                RETRY_INTERVAL_MINUTES,
                RETRY_INTERVAL_MINUTES,
                TimeUnit.MINUTES
        );
        logger.info("Pending note retry scheduler started (interval: " + RETRY_INTERVAL_MINUTES + " min)");
    }

    /**
     * WebSocket 重连成功时调用，立即触发一次重试
     */
    public void onNetworkRestored() {
        if (localCache.getPendingNoteCount() > 0) {
            logger.info("Network restored, triggering pending note retry");
            CompletableFuture.runAsync(this::retryPendingNotes);
        }
    }

    /**
     * 逐条重试发送 pending notes（按 created_at 顺序）
     */
    private void retryPendingNotes() {
        if (retrying) return;
        retrying = true;

        try {
            List<LocalCacheService.PendingNoteData> pendingNotes = localCache.getPendingNotes();
            if (pendingNotes.isEmpty()) return;

            logger.info("Retrying " + pendingNotes.size() + " pending notes...");
            ApiServiceV2 apiService = ServiceManager.getInstance().getApiService();

            for (LocalCacheService.PendingNoteData note : pendingNotes) {
                try {
                    ApiServiceV2.ApiResult result = apiService.postNote(
                            note.content, note.channel, note.createdAt
                    ).get(30, TimeUnit.SECONDS);

                    if (result.success()) {
                        localCache.deletePendingNote(note.id);
                        logger.info("Pending note sent successfully, id=" + note.id);
                    } else {
                        logger.warning("Pending note send failed, id=" + note.id + ": " + result.message());
                        break; // 失败后停止，等下次重试
                    }
                } catch (TimeoutException e) {
                    logger.warning("Pending note send timeout, id=" + note.id);
                    break;
                } catch (Exception e) {
                    logger.warning("Pending note send error, id=" + note.id + ": " + e.getMessage());
                    break;
                }
            }
        } finally {
            retrying = false;
        }
    }

    /**
     * 检查网络是否可用（基于 WebSocket 连接状态）
     */
    public boolean isNetworkAvailable() {
        return ServiceManager.getInstance().getWebSocketService().isConnected();
    }

    public void shutdown() {
        if (retryScheduler != null && !retryScheduler.isShutdown()) {
            retryScheduler.shutdownNow();
            logger.info("Pending note retry scheduler stopped");
        }
    }
}

File: keenotes-android/app/src/main/java/cn/keevol/keenotes/network/PendingNoteService.kt
package cn.keevol.keenotes.network

import android.util.Log
import cn.keevol.keenotes.data.dao.PendingNoteDao
import cn.keevol.keenotes.data.entity.PendingNote
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow

/**
 * 离线暂存笔记的调度服务
 * - 暂存发送失败的笔记
 * - 定时重试发送（30分钟间隔）
 * - WebSocket 重连时立即触发重试
 */
class PendingNoteService(
    private val pendingNoteDao: PendingNoteDao,
    private val apiService: ApiService,
    private val webSocketService: WebSocketService
) {
    companion object {
        private const val TAG = "PendingNoteService"
        private const val RETRY_INTERVAL_MS = 30L * 60 * 1000 // 30 minutes
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var retryJob: Job? = null
    private var isRetrying = false

    val pendingCountFlow: Flow<Int> = pendingNoteDao.getCountFlow()
    val pendingNotesFlow: Flow<List<PendingNote>> = pendingNoteDao.getAllFlow()

    fun startRetryScheduler() {
        if (retryJob != null) return
        retryJob = scope.launch {
            while (isActive) {
                delay(RETRY_INTERVAL_MS)
                retryPendingNotes()
            }
        }
        Log.i(TAG, "Retry scheduler started (interval: 30 min)")

        // WebSocket 重连成功时触发重试
        scope.launch {
            webSocketService.connectionState.collect { state ->
                if (state == WebSocketService.ConnectionState.CONNECTED) {
                    val notes = pendingNoteDao.getAll()
                    if (notes.isNotEmpty()) {
                        Log.i(TAG, "Network restored, retrying ${notes.size} pending notes")
                        retryPendingNotes()
                    }
                }
            }
        }
    }

    /** 暂存一条笔记到本地 */
    fun savePendingNote(content: String, channel: String = "mobile-android") {
        scope.launch {
            try {
                val preparedNote = apiService.prepareNote(content, channel)
                pendingNoteDao.insert(preparedNote.toPendingNote())
                Log.i(TAG, "Note saved to pending")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to save pending note: ${e.message}", e)
            }
        }
    }

    fun savePendingNote(preparedNote: PreparedNote) {
        scope.launch {
            try {
                pendingNoteDao.insert(preparedNote.toPendingNote())
                Log.i(TAG, "Prepared note saved to pending")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to save prepared pending note: ${e.message}", e)
            }
        }
    }

    /** 逐条重试发送 pending notes */
    private suspend fun retryPendingNotes() {
        if (isRetrying) return
        isRetrying = true

        try {
            val pendingNotes = pendingNoteDao.getAll()
            if (pendingNotes.isEmpty()) return

            Log.i(TAG, "Retrying ${pendingNotes.size} pending notes...")

            for (note in pendingNotes) {
                val result = if (!note.encryptedContent.isNullOrEmpty() && !note.requestId.isNullOrEmpty()) {
                    apiService.postPreparedNote(
                        PreparedNote(
                            content = note.content,
                            encryptedContent = note.encryptedContent,
                            channel = note.channel,
                            createdAt = note.createdAt,
                            requestId = note.requestId
                        )
                    )
                } else {
                    apiService.postNote(note.content)
                }
                if (result.success) {
                    pendingNoteDao.deleteById(note.id)
                    Log.i(TAG, "Pending note sent, id=${note.id}")
                } else {
                    Log.w(TAG, "Retry failed: ${result.message}, stopping")
                    if (result.networkError) {
                        webSocketService.markConnectionSuspect("pending-retry-network-error")
                    }
                    break
                }
            }
        } finally {
            isRetrying = false
        }
    }

    /** 检查网络是否可用 */
    fun isNetworkAvailable(): Boolean {
        return webSocketService.connectionState.value == WebSocketService.ConnectionState.CONNECTED
    }

    fun shutdown() {
        retryJob?.cancel()
        scope.cancel()
    }

    private fun PreparedNote.toPendingNote(): PendingNote {
        return PendingNote(
            content = content,
            channel = channel,
            createdAt = createdAt,
            encryptedContent = encryptedContent,
            requestId = requestId
        )
    }
}
